sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c

```diff
@@ -0,0 +1,170 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_pinmux.h"
+#include "sw/device/lib/dif/dif_pwrmgr.h"
+#include "sw/device/lib/dif/dif_rstmgr.h"
+#include "sw/device/lib/dif/dif_sysrst_ctrl.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/pwrmgr_testutils.h"
+#include "sw/device/lib/testing/rstmgr_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+// Test phase and expected values written by testbench.
+static volatile const uint8_t kTestPhase = 0;
+
+static dif_pwrmgr_t pwrmgr;
+static dif_rstmgr_t rstmgr;
+static dif_pinmux_t pinmux;
+static dif_sysrst_ctrl_t sysrst_ctrl;
+
+// This means 20 aon_clk ticks ~= 20 * 5 us = 100 us
+static const uint16_t debounce_timer = 20;
+
+enum {
+  kNumMioInPads = 3,
+  kNumMioOutPads = 1,
+};
+
+const uint32_t kTestPhaseTimeoutUsec = 500;
+enum {
+  kTestPhaseInit = 0,
+  kTestPhaseDriveZero = 1,
+  kTestPhaseWaitNoWakeup = 2,
+  kTestPhaseGlitchLidOpen = 3,
+  kTestPhaseWaitWakeup = 4,
+  kTestPhaseDone = 5,
+};
+
+static const dif_pinmux_index_t kPeripheralInputs[] = {
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonPwrbIn,
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonAcPresent,
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonLidOpen,
+};
+
+static const dif_pinmux_index_t kInputPads[] = {
+    kTopEarlgreyPinmuxInselIor13,
+    kTopEarlgreyPinmuxInselIoc7,
+    kTopEarlgreyPinmuxInselIoc9,
+};
+
+static const dif_pinmux_index_t kPeripheralOutputs[] = {
+    kTopEarlgreyPinmuxOutselSysrstCtrlAonZ3Wakeup,
+};
+
+static const dif_pinmux_index_t kOutputPads[] = {
+    kTopEarlgreyPinmuxMioOutIob7,
+};
+
+// Sets up the pinmux to assign input and output pads
+// to the sysrst_ctrl peripheral as required.
+static void pinmux_setup(void) {
+  for (int i = 0; i < kNumMioInPads; ++i) {
+    CHECK_DIF_OK(
+        dif_pinmux_input_select(&pinmux, kPeripheralInputs[i], kInputPads[i]));
+  }
+
+  for (int i = 0; i < kNumMioOutPads; ++i) {
+    CHECK_DIF_OK(dif_pinmux_output_select(&pinmux, kOutputPads[i],
+                                          kPeripheralOutputs[i]));
+  }
+}
+
+// Waits for the kTestPhase variable to be changed by a backdoor overwrite
+// from the testbench in chip_sw_<testname>_vseq.sv. This will
+// indicate that the testbench is ready to proceed with the
+// next phase of the test.
```

Comment:
Can you use docstring format for all these functions to enable automatic extraction via doxygen?
```
/** 
 * Waits for the kTestPhase variable to be changed by a backdoor overwrite
 * from the testbench in chip_sw_<testname>_vseq.sv. This will
 * indicate that the testbench is ready to proceed with the
 * next phase of the test.
 */
```